### PR TITLE
MAGN-8514 Add Copy/Paste to the context menu

### PR DESCRIPTION
**Pull request by marimano**
### Purpose

Copy/Paste commands are added into the context menu of a workspace. They are visible when CanExecute() is true.
Workspace implements Paste command in order to pass a location where clipboard content should be pasted to. 

Ctrl+V pastes clipboard content into top left worksapce corner with an offset as it used to be.

### Screenshots

There is nothing to copy or paste
nothing_copy_paste

There is a node to copy, but nothing has been copied yet
copy_allowed

Pasting is allowed
paste_allowed

There is a node to copy and something has been copied
copy_paste

### Reviewers

@mjkkirschner

### CODE REVIEW

**Comment by Racel** [comment]
@mmanoylo - What is the state of the context menu when something is in the copy cache (clipboard) and something else is also selected? In this situation, both 'Copy' and 'Paste' should appear in the context menu simultaneously. Let me know what you think. If this is already working, please add an image of it to the review.

**Comment by marimano** [comment]
@Racel yes, it is working as you discribed. I have updated the PR description with a screenshot for this case

**Comment by Racel** [comment]
thanks, @mmanoylo!

File: src/DynamoCore/Models/DynamoModel.cs [hunk 1]
```diff
@@ -1546,10 +1546,30 @@ public void Copy()
...
+            var y = workspaceY + CurrentWorkspace.CurrentPasteOffset;
+            var targetPoint = new Point2D(x, y);
+            
+            Paste(targetPoint);
+        }
+
+        /// <summary>
+        ///     Paste ISelectable objects from the clipboard to the workspace at specified point.
+        /// </summary>
+        public void Paste(Point2D targetPoint)
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs]: as this is a public method please add params tag and description of input

File: src/DynamoCore/Models/DynamoModel.cs [hunk 2]
```diff
@@ -1546,10 +1546,30 @@ public void Copy()
         }
 
         /// <summary>
-        ///     Paste ISelectable objects from the clipboard to the workspace.
+        ///     Paste ISelectable objects from the clipboard to the workspace in top left corner
         /// </summary>
         public void Paste()
         {
+            // Move all of the notes and nodes such that they are aligned with
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs]: just curious? is there a reason we are keeping this method around and not always creating the paste location near the mouse position?

File: src/DynamoCoreWpf/Commands/WorkspaceCommands.cs [hunk 3]
```diff
@@ -206,6 +218,27 @@ public DelegateCommand ShowInCanvasSearchCommand
...
+        {
+            get { return DynamoViewModel.CanPaste(null); }
+        }
+
+        public bool CanCopyOrPaste
+        {
+            get { return CanCopy || CanPaste; }
+        }
+
+        private void Paste(object param)
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Commands/WorkspaceCommands.cs]: I think this method should go on workspaceViewModel like the other methods here to keep the structure parallel. (this is not a property for command data binding)

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 4]
```diff
@@ -447,6 +447,8 @@ protected DynamoViewModel(StartConfiguration startConfiguration)
 
             this.BrandingResourceProvider = startConfiguration.BrandingResourceProvider ?? new DefaultBrandingResourceProvider();
 
+            InitializeDelegateCommands();
+
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: did ``InitializeDelegateCommands`` need to be moved here, if the order is important please add a comment specifying

**Comment by mjkkirschner** [comment]
@marimano just a few small things and some questions.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 2 above)
Comment (marimano) [inline src/DynamoCore/Models/DynamoModel.cs]: when a user press ctrl+v, his mouse position can be anywhere

**Comment by marimano** [comment]
Hi @mjkkirschner, all notes are taken into account in the PR. Please take a look again

**Comment by mjkkirschner** [comment]
@marimano this looks good to me, but it now has some merge conflicts, can you fix them and then I will merge it.

**Comment by marimano** [comment]
Hi @mjkkirschner, it interfered with [another PR](https://github.com/DynamoDS/Dynamo/pull/5361), so I had to change that code in order to compute the point where selection should be pasted.
Both functionalities work in the same way as it was originaly (CTRL+V pasting occurs as described [here](https://github.com/DynamoDS/Dynamo/pull/5361), pasting by mouse puts selection at mouse position)
Please take a look

File: src/DynamoCore/Models/DynamoModel.cs [hunk 5]
```diff
@@ -1548,10 +1548,40 @@ public void Copy()
...
-        ///     Paste ISelectable objects from the clipboard to the workspace.
+        ///     Paste ISelectable objects from the clipboard to the workspace 
+        /// so that the nodes appear in their original location with a slight offset
         /// </summary>
         public void Paste()
         {
+            // Provide a small offset when pasting so duplicate pastes aren't directly on top of each other
+            CurrentWorkspace.IncrementPasteOffset();
+
+            var locatebleModels = ClipBoard.Where(model => model is NoteModel || model is NodeModel);
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs]: Won't this get rid of annotation models? (i.e. groups) is there a reason we can't check that the model is not ILocatable instead?

is there a problem with measuring groups in this way to calculate the bounds?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 6]
```diff
@@ -1548,10 +1548,40 @@ public void Copy()
...
+            var orderedItems = locatebleModels.OrderBy(item => item.CenterX + item.CenterY);
+
+            // Search for the rightmost item. It's item with the biggest X, Y coordinates of center.
+            var rightMostItem = orderedItems.Last();
+
+            // Search for the leftmost item. It's item with the smallest X, Y coordinates of center.
+            var leftMostItem = orderedItems.First();
+
+            var shiftX = rightMostItem.X + rightMostItem.Width + CurrentWorkspace.CurrentPasteOffset - leftMostItem.X;
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs]: can you add the comments back here to describe this algorithm

File: src/DynamoCore/Models/DynamoModel.cs (hunk 5 above)
Comment (marimano) [inline src/DynamoCore/Models/DynamoModel.cs]: Annotations did not participate in computing `minXY`, `maxXY`. So I have not included them here

**Comment by mjkkirschner** [comment]
@marimano thanks for this, sorry it got out of sync, I'm going to run this on the self serve CI and merge it if all is good.

**Comment by mjkkirschner** [comment]
merging this in, there were 0 failures on the CI

**Pull request merged**